analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/BinaryBase.cs

```diff
@@ -0,0 +1,296 @@
+            : null;
+
+    private static BoolConstraint BoolConstraintFromBoolAndNullConstraints(
+        BinaryOperatorKind kind,
+        BoolConstraint leftBool,
+        BoolConstraint rightBool,
+        bool leftIsNull,
+        bool rightIsNull) =>
+        kind switch
+        {
+            BinaryOperatorKind.Equals when leftIsNull || rightIsNull => BoolConstraint.False,
+            BinaryOperatorKind.NotEquals when leftIsNull || rightIsNull => BoolConstraint.True,
+            BinaryOperatorKind.Or or BinaryOperatorKind.ConditionalOr when (leftBool ?? rightBool) == BoolConstraint.True => BoolConstraint.True,
+            BinaryOperatorKind.And or BinaryOperatorKind.ConditionalAnd when (leftBool ?? rightBool) == BoolConstraint.False => BoolConstraint.False,
+            _ => null
+        };
+
+    private static NumberConstraint LearnRelationalNumberConstraint(NumberConstraint existingNumber, BinaryOperatorKind kind, NumberConstraint comparedNumber) =>
+        kind switch
+        {
+            BinaryOperatorKind.Equals => NumberConstraint.From(
+                ArithmeticCalculator.BiggestMinimum(comparedNumber, existingNumber),
+                ArithmeticCalculator.SmallestMaximum(comparedNumber, existingNumber)),
+            BinaryOperatorKind.NotEquals when comparedNumber.IsSingleValue && comparedNumber.Min == existingNumber?.Min =>
+                NumberConstraint.From(existingNumber!.Min + 1, existingNumber.Max),
+            BinaryOperatorKind.NotEquals when comparedNumber.IsSingleValue && comparedNumber.Min == existingNumber?.Max =>
+                NumberConstraint.From(existingNumber!.Min, existingNumber.Max - 1),
+            BinaryOperatorKind.GreaterThan when comparedNumber.Min.HasValue =>
+                From(comparedNumber.Min + 1, null, existingNumber),
+            BinaryOperatorKind.GreaterThanOrEqual when comparedNumber.Min.HasValue =>
+                From(comparedNumber.Min, null, existingNumber),
+            BinaryOperatorKind.LessThan when comparedNumber.Max.HasValue =>
+                From(null, comparedNumber.Max - 1, existingNumber),
+            BinaryOperatorKind.LessThanOrEqual when comparedNumber.Max.HasValue =>
+                From(null, comparedNumber.Max, existingNumber),
+            _ => null
+        };
+
+    private static NumberConstraint From(BigInteger? newMin, BigInteger? newMax, NumberConstraint existingNumber)
+    {
+        if (existingNumber is not null)
+        {
+            if (newMin is null || existingNumber.Min > newMin)
+            {
+                newMin = existingNumber.Min;
+            }
+            if (newMax is null || existingNumber.Max < newMax)
+            {
+                newMax = existingNumber.Max;
+            }
+        }
+        return NumberConstraint.From(newMin, newMax);
+    }
+
+    protected static SymbolicConstraint BinaryOperandConstraint<T>(ProgramState state, IOperation leftOperand, IOperation rightOperand) where T : SymbolicConstraint =>
```

Comment:
```suggestion
    protected static T BinaryOperandConstraint<T>(ProgramState state, IOperation leftOperand, IOperation rightOperand) where T : SymbolicConstraint =>
```

Reply:
It will not work because of [this line](https://github.com/SonarSource/sonar-dotnet/blob/41a94a002b1f3d2b80457a502783cc2b191c161d/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/BinaryBase.cs#L90). `ApplyOpposite` returns `SymbolicConstraint` and I'd have to convert it to `T`, which wouldn't look good.